Userland/Applications/Escalator/EscalatorWindow.cpp

```diff
@@ -0,0 +1,158 @@
+            close();
+        }
+    };
+    m_ok_button->set_default(true);
+
+    m_cancel_button = *main_widget.find_descendant_of_type_named<GUI::DialogButton>("cancel_button");
+    m_cancel_button->on_click = [this](auto) {
+        close();
+    };
+
+    m_password_input = *main_widget.find_descendant_of_type_named<GUI::PasswordBox>("password");
+}
+
+GUI::Icon EscalatorWindow::find_icon()
+{
+    String icon_path = m_command[0];
+
+    // Path to executable directly provided.
+    if (icon_path.find('/').has_value())
+        return GUI::FileIconProvider::icon_for_executable(icon_path);
+
+    // Find the application executable on the user's PATH.
+    // If PATH is absent from the environ for whatever reason, fallback on /bin.
+    String path = "/bin";
+    for (size_t i = 0; environ[i]; ++i) {
+        StringView environ_setting { environ[i], strlen(environ[i]) };
+        auto split = environ_setting.split_view('=');
+
+        if (split.size() == 2 && split[0] == "PATH" && split[1].length()) {
+            path = split[1];
+            break;
+        }
+    }
+
+    Vector<StringView> paths = path.split_view(':');
+    for (auto const& path_entry : paths) {
+        String full_path = String::formatted("{}/{}", path_entry, m_command[0]);
+        auto file_or_error = Core::MappedFile::map(full_path);
+        if (!file_or_error.is_error())
+            return GUI::FileIconProvider::icon_for_executable(full_path);
+    }
+
+    return GUI::Icon::default_icon("app-escalator"sv);
+}
+
+ErrorOr<void> EscalatorWindow::check_password()
+{
+    hide();
+
+    String password = m_password_input->text();
+    if (password.is_empty()) {
+        GUI::MessageBox::show_error(this, "Please enter a password."sv);
+        show();
+        return {};
+    }
+
+    // FIXME: PasswordBox really should store its input directly as a SecretString.
+    Core::SecretString password_secret = Core::SecretString::take_ownership(password.to_byte_buffer());
+    if (!m_current_user.authenticate(password_secret)) {
+        GUI::MessageBox::show_error(this, "Incorrect or disabled password."sv);
+        show();
+        return {};
+    }
+
+    // Caller will close Escalator if error is returned.
+    TRY(execute_command());
+    close();
+    return {};
+}
+
+ErrorOr<int> EscalatorWindow::execute_command()
+{
+    String shell = m_current_user.shell();
+
+    // Join all variable args from trailing command.
+    StringBuilder builder;
+    builder.join(' ', m_command);
+    String command = builder.to_string();
+
+    char const* shell_executable = shell.characters();
+    char const* argv[] = { shell_executable, "-c", command.characters(), nullptr };
+
+    // Escalate process privilege to root user.
+    TRY(Core::System::seteuid(0));
+    auto root_user = TRY(Core::Account::from_uid(0));
+    TRY(root_user.login());
+
+    TRY(Core::System::pledge("stdio recvfd sendfd thread cpath rpath wpath unix proc exec"));
```

Comment:
I would expect that this pledge set can be reduced at this point in the program

Reply:
Reduced in latest push